src/Avalonia.Base/Input/GestureRecognizers/VelocityTracker.cs

```diff
@@ -0,0 +1,424 @@
+    /// estimate's [confidence] measures how well the velocity tracker's position
+    /// data fit a straight line, [duration] is the time that elapsed between the
+    /// first and last position sample used to compute the velocity, and [offset]
+    /// is similarly the difference between the first and last positions.
+    ///
+    /// See also:
+    ///
+    ///  * [VelocityTracker], which computes [VelocityEstimate]s.
+    ///  * [Velocity], which encapsulates (just) a velocity vector and provides some
+    ///    useful velocity operations.
+    public record VelocityEstimate(Vector PixelsPerSecond, double Confidence, TimeSpan Duration, Vector Offset);
+
+    internal record struct PointAtTime(bool Valid, Vector Point, TimeSpan Time);
+
+    /// Computes a pointer's velocity based on data from [PointerMoveEvent]s.
+    ///
+    /// The input data is provided by calling [addPosition]. Adding data is cheap.
+    ///
+    /// To obtain a velocity, call [getVelocity] or [getVelocityEstimate]. This will
+    /// compute the velocity based on the data added so far. Only call these when
+    /// you need to use the velocity, as they are comparatively expensive.
+    ///
+    /// The quality of the velocity estimation will be better if more data points
+    /// have been received.
+    public class VelocityTracker
+    {
+        private const int AssumePointerMoveStoppedMilliseconds = 40;
+        private const int HistorySize = 20;
+        private const int HorizonMilliseconds = 100;
+        private const int MinSampleSize = 3;
+        private const double MinFlingVelocity = 50.0; // Logical pixels / second (defined in flutter\lib\src\gesture\constants.dart)
+        private const double MaxFlingVelocity = 8000.0;
+
+        private readonly PointAtTime[] _samples = new PointAtTime[HistorySize];
+        private int _index = 0;
+
+        /// <summary>
+        /// Adds a position as the given time to the tracker.
+        /// </summary>
+        /// <param name="time"></param>
+        /// <param name="position"></param>
+        public void AddPosition(TimeSpan time, Vector position)
+        {
+            _index++;
+            if (_index == HistorySize)
+            {
+                _index = 0;
+            }
+            _samples[_index] = new PointAtTime(true, position, time);
+        }
+
+        /// Returns an estimate of the velocity of the object being tracked by the
+        /// tracker given the current information available to the tracker.
+        ///
+        /// Information is added using [addPosition].
+        ///
+        /// Returns null if there is no data on which to base an estimate.
+        protected virtual VelocityEstimate? GetVelocityEstimate()
+        {
+            double[] x = new double[HistorySize];
+            double[] y = new double[HistorySize];
+            double[] w = new double[HistorySize];
+            double[] time = new double[HistorySize];
```

Comment:
This can be either be allocated on stack (stackalloc) or with reused arrays, as HistorySize isn't changed.